File: android/app/src/main/kotlin/jp/co/bitz/tictactoe/game/states/HumanTurnState.kt
package jp.co.bitz.tictactoe.game.states

import jp.co.bitz.gameplaykit.GKState
import jp.co.bitz.tictactoe.game.TicTacToeMatch
import jp.co.bitz.tictactoe.game.TicTacToeMove
import kotlin.reflect.KClass

/**
 * Waits for the SpriteKit layer (Phase 2) to report a tap via [applyHumanMove]. See
 * docs/GAME_DESIGN.md's "Turn flow" section.
 */
public class HumanTurnState(
    private val match: TicTacToeMatch,
) : GKState() {
    /**
     * Applies a tap at [cellIndex] if it's a legal move for the active player, advancing to
     * [TurnEndState]. Returns whether the move was applied.
     */
    public fun applyHumanMove(cellIndex: Int): Boolean {
        val move =
            match.gameModel.activePlayer
                ?.let { match.gameModel.gameModelUpdates(it) }
                ?.filterIsInstance<TicTacToeMove>()
                ?.firstOrNull { it.cellIndex == cellIndex }
        return if (move != null) {
            match.gameModel.apply(move)
            stateMachine?.enter<TurnEndState>() ?: false
        } else {
            false
        }
    }

    override fun isValidNextState(stateClass: KClass<out GKState>): Boolean = stateClass == TurnEndState::class
}

File: ios/TicTacToe/Game/States/HumanTurnState.swift
import GameplayKit

/// Waits for the SpriteKit layer (Phase 2) to report a tap via `applyHumanMove(at:)`. See
/// `docs/GAME_DESIGN.md`'s "Turn flow" section.
final class HumanTurnState: GKState {
    unowned let match: TicTacToeMatch

    init(match: TicTacToeMatch) {
        self.match = match
    }

    /// Applies a tap at `cellIndex` if it's a legal move for the active player, advancing to
    /// `TurnEndState`. Returns whether the move was applied.
    @discardableResult
    func applyHumanMove(at cellIndex: Int) -> Bool {
        guard let activePlayer = match.gameModel.activePlayer,
              let updates = match.gameModel.gameModelUpdates(for: activePlayer) as? [TicTacToeMove],
              let move = updates.first(where: { $0.cellIndex == cellIndex })
        else {
            return false
        }
        match.gameModel.apply(move)
        return stateMachine?.enter(TurnEndState.self) ?? false
    }

    override func isValidNextState(_ stateClass: AnyClass) -> Bool {
        stateClass == TurnEndState.self
    }
}
